### ISSUE TITLE
Fails when the query uses populate for referenced sub-docs

### ISSUE THREAD

**Issue by edwardsph** [vkarpov15/mongoose-lean-virtuals#42]
I've been using this library for a while with just {virtuals: true} and found it really helpful but I need to limit the number of virtuals being returned so attempted to replace `true` with an array of names. It almost works but I was getting an error when using this in a query that populates referenced docs. The line below fails because applyGetters is called against an undefined value:
```
cur[sp[sp.length - 1]] = schema.virtuals[virtual].applyGetters(cur[sp[sp.length - 1]], doc);
```
Simply testing to see if `schema.virtuals[virtual]` exists before this line seems to have worked. One concern I have is that the virtuals provided in the list are not specific to any particular schema, The code will give you virtuals from every sub-schema where a virtual matches the name. That is fine once understood but it would be really nice to be able to be more specific. For example if you have a student record with a contact record you might want to specify virtuals as `['currentStatus', 'contact.fullName']` to get the first from the top level schema and the second from the specific child schema.

**Comment by vkarpov15** [comment]
I'm unable to repro this issue, can you please modify the below script to demonstrate your issue?

```javascript
'use strict';

const mongoose = require('mongoose');

mongoose.set('debug', true);

run().catch(err => console.log(err));

async function run() {
  await mongoose.connect('mongodb://localhost:27017/test', {
    useNewUrlParser: true,
    useUnifiedTopology: true
  });

  await mongoose.connection.dropDatabase();

  const contactSchema = new mongoose.Schema({
    firstName: String,
    lastName: String
  });
  contactSchema.virtual('name').get(function() {
    return `${this.firstName} ${this.lastName}`;
  });

    const schema = new mongoose.Schema({
    contact: contactSchema
  });
  schema.virtual('test').get(() => 42);
  schema.plugin(require('mongoose-lean-virtuals'));
  const Test = mongoose.model('Test', schema);

  await Test.create({ contact: { firstName: 'test', lastName: 'test' } });

  const doc = await Test.findOne().lean({ virtuals: ['contact.name'] });
  console.log(doc);
}
```

**Comment by edwardsph** [comment]
I've reproduced the problem I was trying to describe - the difference is that I am using 2 collections with populate to include the referenced document.

```
'use strict';

const mongoose = require('mongoose');

mongoose.set('debug', true);

run().then(() => {process.exit(0)}).catch(err => console.log(err));

async function run() {
  await mongoose.connect('mongodb://localhost:27017/test', {
    useNewUrlParser: true,
    useUnifiedTopology: true
  });

  await mongoose.connection.dropDatabase();

  const contactSchema = new mongoose.Schema({
    firstName: String,
    lastName: String
  });
  contactSchema.virtual('name').get(function() {
    return `${this.firstName} ${this.lastName}`;
  });
  contactSchema.plugin(require('./local-lean-virtuals'));
  // contactSchema.plugin(require('mongoose-lean-virtuals'));
  const Contact = mongoose.model('Contact', contactSchema);
  const contact = await Contact.create({ firstName: 'test', lastName: 'test' });

  const schema = new mongoose.Schema({
    contact: {type: mongoose.Schema.Types.ObjectId, ref: 'Contact'},
  });
  schema.virtual('test').get(() => 42);
  schema.plugin(require('./local-lean-virtuals'));
  // schema.plugin(require('mongoose-lean-virtuals'));

  const Test = mongoose.model('Test', schema);
  await Test.create({ contact: contact._id });

  // const doc = await Test.findOne().lean();
  const doc = await Test.findOne().populate('contact').lean();
  console.log('Populate without virtuals: ', doc);

  const doc2 = await Test.findOne().populate('contact').lean({ virtuals: ['test', 'name'] });
  console.log('Populate with virtuals: ', doc2);
}
```
I have got around the issue by replacing your module with "local-lean-virtuals" - the only difference is the `if` statement at the end:
```
    if (schema.virtuals[virtual]) {
      cur[sp[sp.length - 1]] = schema.virtuals[virtual].applyGetters(cur[sp[sp.length - 1]], doc);
    }
```
The other change is that I found the same list of virtual fields requested is passed to each schema so at that point the nested reference doesn't work so the query need to become:
```
  const doc2 = await Test.findOne().populate('contact').lean({ virtuals: ['test', 'name'] });
```

In an ideal world we would be able to call:
```
  const doc2 = await Test.findOne().populate('contact').lean({ virtuals: ['test', 'contact.name'] });
```
and the module would work out how to apply the virtuals to referenced documents and wouldn't attempt to get the virtual `test` from the contact schema.

I hope that makes sense. I can put the project in github if you like but all the code is outlined here. It would be great of there was a way to extend the functionality to support this as it would be really useful.

**Comment by vkarpov15** [comment]
The issue is that there's no `name` virtual on the `Test` model, so that's why you're getting this error.

Can you please explain what you mean by "the module would work out how to apply the virtuals to referenced documents and wouldn't attempt to get the virtual test from the contact schema"? I don't understand. Perhaps a code sample might help.

**Comment by edwardsph** [comment]
Sorry for the slow response - I just started a new job. I've produced an example that shows how embedded sub-docs (Organisation) are treated differently to referenced sub-docs (Contact). In incorporated your fix to my local copy of mongoose-lean-virtuals. Both Contact and Organisation have a virtual field named "name" but the second query `{ virtuals: ['test', 'name'] }` only returns test and contact.name. The third query `{ virtuals: ['test', 'name', 'organisation.name'] }` returns test, contact.name and organisation.name. The things to note are:
1. The result doc embeds contact and organisation in the same way and both have a name field
2. When referencing virtual fields of a referenced doc you need to use the bare field name
3. If there was a second referenced schema (e.g. Friend) that also had a virtual field called name then you could not say you wanted one without the other. Adding `name` to the virtuals list would return Contact.name and Friend.name
4. When using an embedded doc, the fact that it is nested becomes significant and to fetch the name field you have to select it with `organisation.name`
5. This means that currently the way you reference fields of sub-docs changes dependant on whether they are referenced or embedded

Ideally every field would be specifically referenced as it appears in the output document since that matches the overall model being queried. So, the virtuals list would be `{ virtuals: ['test', 'contact.name', 'organisation.name'] }`. When the virtuals list is passed to the contact model, only fields with the contact discriminator would be extracted similar to how the organisation discriminator is handled at present.

I don't know how easy that is to achieve but it would be great to have this level of consistency and the ability to have full control over the selecting of fields.

```js
'use strict';

const mongoose = require('mongoose');

mongoose.set('debug', true);

run().then(() => {process.exit(0)}).catch(err => console.log(err));

async function run() {
  await mongoose.connect('mongodb://localhost:27017/test', {
    useNewUrlParser: true,
    useUnifiedTopology: true
  });
  await mongoose.connection.dropDatabase();

  // Setup Contact model with virtual "name"
  const contactSchema = new mongoose.Schema({
    firstName: String,
    lastName: String
  });
  contactSchema.virtual('name').get(function() {
    return `${this.firstName} ${this.lastName}`;
  });
  contactSchema.plugin(require('./local-lean-virtuals'));
  // contactSchema.plugin(require('mongoose-lean-virtuals'));
  const Contact = mongoose.model('Contact', contactSchema);
  const contact = await Contact.create({ firstName: 'First', lastName: 'Last' });

  // Setup Organisation model with virtual "name"
  const organisationSchema = new mongoose.Schema({
    orgName: String
  });
  organisationSchema.virtual('name').get(function() {
    return `Organisation name: ${this.orgName}`;
  });
  organisationSchema.plugin(require('./local-lean-virtuals'));
  // organisationSchema.plugin(require('mongoose-lean-virtuals'));
  const Organisation = mongoose.model('Contact', contactSchema);
  const organisation = await Contact.create({ firstName: 'test', lastName: 'test' });

  // Setup test model with referenced Contact and embedded Organisation and virtual "test"
  const schema = new mongoose.Schema({
    contact: {type: mongoose.Schema.Types.ObjectId, ref: 'Contact'},
    organisation: organisationSchema
  });
  schema.virtual('test').get(() => 42);
  schema.plugin(require('./local-lean-virtuals'));
  // schema.plugin(require('mongoose-lean-virtuals'));
  const Test = mongoose.model('Test', schema);
  await Test.create({
    contact: contact._id,
    organisation: {
      orgName: 'Company'
    }
  });

  const doc = await Test.findOne().populate('contact').lean();
  console.log('Populate without virtuals (contact.name and organisation.name not present): ', doc);

  const doc2 = await Test.findOne().populate('contact').lean({ virtuals: ['test', 'name'] });
  console.log('Populate with virtuals (test and contact.name present but not organisation.name): ', doc2);

  const doc3 = await Test.findOne().populate('contact').lean({ virtuals: ['test', 'name', 'organisation.name'] });
  console.log('Populate with virtuals (test, contact.name and organisation.name present): ', doc3);
}
```

**Comment by vkarpov15** [comment]
Fix will be in Mongoose 5.11.9 :+1:

**Comment by edwardsph** [comment]
I just tested this and whilst it gets further there still seems to be a problem when nested schemas and referenced schemas are both used. I've just set up PR #48 with a test that demonstrates the issue.
